### ISSUE TITLE
`test` task and pre-compile issue

### ISSUE THREAD

**Issue by fvictorio** [NomicFoundation/hardhat#1864]
Reproduction steps:

1. `git clone https://github.com/fvictorio/nexusmutual`
2. Install dependencies and run tests twice.
3. Notice that the error message is different the second time

Another way of doing this is running `hh clean && hh test` vs `hh clean && hh compile && hh test`

The reason might be related to this:

```js
task('test', async (_, hre, runSuper) => {
  hre.accounts = await hre.web3.eth.getAccounts();
  const testFiles = _.testFiles.length ? _.testFiles : ['./test/index.js'];
  await runSuper({ testFiles });
});
```

**Comment by sisco0** [comment]
Output logs for the two tests runs are included next.

# First run
---
  Nexus Mutual
    UNIT TESTS
      TokenController
        removeEmptyReason
          1) reverts when index is out of bounds


  0 passing (357ms)
  1 failing

  1) Nexus Mutual
       UNIT TESTS
         TokenController
           removeEmptyReason
             reverts when index is out of bounds:
     Error: Transaction reverted without a reason string
      at <UnrecognizedContract>.<unknown> (0x5fc8d32690cc91d4c39d9d3abcbd16989f875707)
      at TruffleContract.removeEmptyReason (node_modules/@nomiclabs/truffle-contract/lib/execute.js:169:26)
      at Context.<anonymous> (test/unit/TokenController/removeEmptyReason.js:39:27)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
---
# Second run
---
  Nexus Mutual
    UNIT TESTS
      TokenController
        removeEmptyReason
          1) reverts when index is out of bounds


  0 passing (369ms)
  1 failing

  1) Nexus Mutual
       UNIT TESTS
         TokenController
           removeEmptyReason
             reverts when index is out of bounds:
     Error: VM Exception while processing transaction: invalid opcode
      at TokenController._removeEmptyReason (contracts/modules/token/TokenController.sol:658)
      at TokenController.removeEmptyReason (contracts/modules/token/TokenController.sol:636)
      at TruffleContract.removeEmptyReason (node_modules/@nomiclabs/truffle-contract/lib/execute.js:169:26)
      at Context.<anonymous> (test/unit/TokenController/removeEmptyReason.js:39:27)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)

**Comment by fvictorio** [comment]
@sisco0 you've seem to have fixed something in the reproduction example, but the problem has to be in Hardhat: even if the test is wrong, the fact that the results are inconsistent (in one scenario de test case passes but in the other it doesn't) suggest that there might be a bug on Hardhat's side.

**Comment by alcuadrado** [comment]
This error is the same one that I mentioned as (1) [here](https://github.com/nomiclabs/hardhat/issues/1996#issuecomment-1006110954), and I've already fixed it. I'll submit a PR tomorrow.

**Comment by feuGeneA** [comment]
Fixed via #2218
